### PR TITLE
fix(@aws-amplify/datastore): strictly define null vs undefined behavior on models

### PR DESCRIPTION
**Pull request by svidgen**
#### Resubmitted

Originally submitted as #10915, which caused failing integ tests, due to `owner: null` being submitted in mutations when the the `owner` field was made explicit. AppSync rejects `owner: null` as "Unauthorized."

This updated PR omits the owner field in mutations (or custom owner field per model `attributes.auth`) when the value would be `null` or `undefined`.

Added tests for `owner` handling are in [connectivityHandling.test.ts:102-168](https://github.com/aws-amplify/amplify-js/pull/10932/files#diff-68fad1b028a70ed30e99e616e38d45ec29a2993942226f28ff94cdb241bb3b25R102-R168), with owner field validation occurring in [the graphql fake in helper.ts](https://github.com/aws-amplify/amplify-js/pull/10932/files#diff-d80854c979fb9c08302352685e794dcd2cd3c7a4d29234d90b7757061d14ca5eR779-R781).

The fix is for `owner` handling is in `mutation.ts` on [lines 487-490](https://github.com/aws-amplify/amplify-js/pull/10932/files#diff-caf2490eda4f0fb658881cae74f9e9b5db5e70363b8d0a737bd7bdaca51e053aR487-R490) and [lines 526-529](https://github.com/aws-amplify/amplify-js/pull/10932/files#diff-caf2490eda4f0fb658881cae74f9e9b5db5e70363b8d0a737bd7bdaca51e053aR526-R529).

You can also see the bulk of the `owner` field fix and tests together by looking at the [2nd commit in the PR](https://github.com/aws-amplify/amplify-js/pull/10932/commits/bc756645a40317454cdfa6e2e94c4952a93efe09).

#### Description of changes

DataStore currently treats nullable model fields different before/after synchronization, which can lead to unexpected results when operating against or querying these instances. For example:

Given:

```graphql
type Model @model {
  id: ID!
  value: String!
  otherValue: String
}
```

If we save an instance:

```js
const saved = await DataStore.save(new Model({
    value: 'some value'
    /* otherValue: 'oh nevermind!' */
}));
```

We'll have this returned:

```json
{
  "value": "some value",
  "id": "a9f6e211-1d91-45c6-8fbe-46173fb9f09b"
}
```

If we query by ID *before* the instance is synchronized to the cloud, we see the same thing. And, if we query for models where `m.otherValue.eq(null)`, the model is *not* in the result set, because the field is actually `undefined`.

However, after synchronization, the the cloud storage provider (e.g., AppSync) will set optional fields to `null`:

```json
{
  "id": "a9f6e211-1d91-45c6-8fbe-46173fb9f09b",
  "value": "unique instance 0.21662066323138118",

  // LOOK!
  "otherValue": null, 

  "createdAt": "2023-01-27T18:02:40.093Z",
  "updatedAt": "2023-01-27T18:02:40.093Z",
  "owner": "test",
  "_version": 1,
  "_lastChangedAt": 1674842560119,
  "_deleted": null
}
```

The model will then appear in the result set if we search for `m => m.otherValue.eq(null)`.

This is an unintended behavior that customers should not have to work around. This PR normalizes that behavior so that optional fields default to `null` when unspecified or specified as `undefined`. (`undefined` can't actually be persisted in cloud storage.)

#### Description of how you validated changes

1. Added and updated unit tests.
2. Verified in sample apps.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by svidgen** [comment]
Spot-checked failing integ tests. They are now passing locally.

image

File: packages/datastore/src/sync/processors/mutation.ts [hunk 1]
```diff
@@ -484,6 +484,11 @@ class MutationProcessor {
 		const modelDefinition = this.schema.namespaces[namespaceName].models[model];
 		const { primaryKey } = this.schema.namespaces[namespaceName].keys![model];
 
+		const auth = modelDefinition.attributes?.find(a => a.type === 'auth');
+		const ownerField = auth?.properties?.rules.find(
+			rule => rule.ownerField
+		)?.ownerField;
```
Comment (iartemiev) [inline packages/datastore/src/sync/processors/mutation.ts:490]: Seems like this only handles a single owner field, but models can have multiple owner fields configured, e.g. 
```gql
type Post @model @auth(rules: [ 
  { allow: owner }, 
  { allow: owner, ownerField: editor, operations: [read, update] } 
]) {
  id: ID!
  owner: String
  editor: String
}
```
I think we should do `auth?.properties?.rules.filter` and then iterate over the resulting array below

File: packages/datastore/package.json [hunk 2]
```diff
@@ -24,8 +24,7 @@
 		"build:cjs:watch": "node ./build es5 --watch",
 		"build:esm:watch": "rimraf lib-esm && node ./build es6 --watch",
 		"build": "yarn clean && yarn build:esm && npm run build:cjs",
-		"clean": "npm run clean:size && rimraf lib-esm lib dist",
-		"clean:size": "rimraf dual-publish-tmp tmp*",
```
Comment (dpilch) [inline packages/datastore/package.json:28]: Accidental removal?

Comment (svidgen) [reply packages/datastore/package.json:28]: Ah yes. Bad merge. Fixing...

File: packages/datastore/src/sync/processors/mutation.ts (hunk 1 above)
Comment (svidgen) [reply packages/datastore/src/sync/processors/mutation.ts:490]: That's a pretty important gap to call out! 😅 Thank-you!

Comment (svidgen) [reply packages/datastore/src/sync/processors/mutation.ts]: Fixed.

**Review by dpilch — APPROVED** [review]

**Review by iartemiev — APPROVED** [review]

**Pull request merged**